### PR TITLE
[SPARK-46512][CORE] Optimize shuffle reading when both sort and combine are used.

### PR DESCRIPTION
**Pull request by zhengchenyu**
### What changes were proposed in this pull request?

After the shuffle reader obtains the block, it will first perform a combine operation, and then perform a sort operation. It is known that both combine and sort may generate temporary files, so the performance may be poor when both sort and combine are used. In fact, combine operations can be performed during the sort process, and we can avoid the combine spill file.
See  https://issues.apache.org/jira/browse/SPARK-46512 for details

### Why are the changes needed?

Reduce the number of spills to disk when both sort and combine are used

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

spark examples run in local and yarn mode.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Review by Ngone51 — COMMENTED** [review]
I'm trying to understand what's exact difference before and after the change. Before the change, we do iterator combination (whether combiners or values) first and then sort the combined iterator. After the change, the combination work seems just to be moved into `ExternalSorter` only. However, `ExternalSorter` also do combination first (`insertAllAndUpdateMetrics()`) and then sort the combined iterator (when call `iterator`). `ExternalSorter` doesn't really do combination and sort at the same time. So it looks like there's no big difference. Can you help elaborate more?

**Comment by zhengchenyu** [comment]
@Ngone51 
Thanks for you review! The combine is the key problem. 
I still combine, but combine in ExternalSorter will never trigger extra spill. Combine happen in insertAllAndUpdateMetrics and merge:
(1) combine when insert, just read the exist key and combine the value, do it in memory. 
(2) combine when merge, we sort the combined iterator, combine util next key occur, also do it in memory.
The two processes are not additional operations and are originally performed in the ExternalSorter.

Before the change, We use ExternalAppendOnlyMap to combine. If the memory is over threshold, will spill to disk. Then we use ExternalSorter to sort. If the memory is over threshold, will spill to disk. **It means we may spill twice(ExternalAppendOnlyMap and ExternalSorter), if shuffle data is huge.**

We know that we can combine during the sorting process. 
After the change, we will use ExternalSorter, combine occurs during insert and merge processes. So there is no need to call ExternalAppendOnlyMap to combine. **It means only spill once (ExternalSorter).**

Spill is time consuming. For sort and combine shuffle, when shuffle date is huge, this PR will save time。

**Comment by Ngone51** [comment]
> but combine in ExternalSorter will never trigger extra spill.

It could be triggered by `ExternalSorter#maybeSpillCollection`?

**Comment by Ngone51** [comment]
It looks like `ExternalAppendOnlyMap` uses the key for combine:
```
dep.aggregator.get.combineCombinersByKey(combinedKeyValuesIterator, context)
```

 but `ExternalSorter` uses partition id:

``` 
map.changeValue((actualPartitioner.getPartition(kv._1), kv._1), update)
```

Do you think this would lead to the different result?

**Comment by zhengchenyu** [comment]
> > but combine in ExternalSorter will never trigger extra spill.
> 
> It could be triggered by `ExternalSorter#maybeSpillCollection`?

@Ngone51 Thanks for your reply!

Yes, It could be triggered by `ExternalSorter#maybeSpillCollection`. I mean that the Extra spill is trigger by `ExternalAppendOnlyMap#spill`

Before the change, both`ExternalAppendOnlyMap#spill` and `ExternalSorter#maybeSpillCollection` will be spill. After the change, only `ExternalSorter#maybeSpillCollection` will be spill.

I created a huge amount of data, and I found the following logs:

```
23/12/26 16:55:52 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 522.3 MiB to disk (1 time so far)
23/12/26 16:56:09 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 629.8 MiB to disk (2 times so far)
23/12/26 16:56:40 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 491.5 MiB to disk (3 times so far)
23/12/26 16:56:58 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 553.0 MiB to disk (4 times so far)
23/12/26 16:57:16 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 442.9 MiB to disk (5 times so far)
23/12/26 16:57:47 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 576.1 MiB to disk (6 times so far)
23/12/26 16:58:04 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 540.2 MiB to disk (7 times so far)
23/12/26 16:58:22 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 686.3 MiB to disk (8 times so far)
23/12/26 16:58:52 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 540.2 MiB to disk (9 times so far)
23/12/26 17:00:06 INFO ExternalAppendOnlyMap: Task 5 force spilling in-memory map to disk and it will release 942.2 MiB memory
23/12/26 17:00:27 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (1 time so far)
23/12/26 17:01:24 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (2 times so far)
23/12/26 17:02:24 INFO ExternalSorter: Thread 48 spilling in-memory map of 959.1 MiB to disk (3 times so far)
23/12/26 17:03:23 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (4 times so far)
23/12/26 17:04:30 INFO ExternalSorter: Thread 48 spilling in-memory map of 948.8 MiB to disk (5 times so far)
23/12/26 17:05:24 INFO ExternalSorter: Thread 48 spilling in-memory map of 943.7 MiB to disk (6 times so far)
23/12/26 17:06:25 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (7 times so far)
23/12/26 17:07:08 INFO ExternalSorter: Thread 48 spilling in-memory map of 959.1 MiB to disk (8 times so far)
23/12/26 17:07:58 INFO ExternalSorter: Thread 48 spilling in-memory map of 959.1 MiB to disk (9 times so far) 
```

After the change, we will only found ExternalSorter spill log.

**Comment by zhengchenyu** [comment]
> It looks like `ExternalAppendOnlyMap` uses the key for combine:
> 
> ```
> dep.aggregator.get.combineCombinersByKey(combinedKeyValuesIterator, context)
> ```
> 
> but `ExternalSorter` uses partition id:
> 
> ```
> map.changeValue((actualPartitioner.getPartition(kv._1), kv._1), update)
> ```
> 
> Do you think this would lead to the different result?

@Ngone51 
It seems that the key for combine is Tuple2(partitionid, key).

**Comment by Ngone51** [comment]
I see @zhengchenyu

**Comment by mridulm** [comment]
Thanks for looking into this @Ngone51 !
Will take a look over this PR later this week/next - sorry for the delay, was away on a vacation.
+CC @otterc as well.

File: core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala [hunk 1]
```diff
@@ -111,31 +111,50 @@ private[spark] class BlockStoreShuffleReader[K, C](
...
         // We don't know the value type, but also don't care -- the dependency *should*
         // have made sure its compatible w/ this aggregator, which will convert the value
         // type to the combined type C
-        val keyValuesIterator = interruptibleIter.asInstanceOf[Iterator[(K, Nothing)]]
+        val keyValuesIterator = sortedIter.asInstanceOf[Iterator[(K, Nothing)]]
         dep.aggregator.get.combineValuesByKey(keyValuesIterator, context)
       }
     } else {
-      interruptibleIter.asInstanceOf[Iterator[Product2[K, C]]]
-    }
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala:128]: We can drop the `aggregated` flag.

In other words, we can make this simply:

```
val resultIter = {
  if (dep.keyOrdering.isDefined) {
    // whatever is in dep.keyOrdering == match `Some(keyOrd: Ordering[K])` clause.
  } else if (dep.aggregator.isDefined) {

    // whatever is in "if (!aggregated && dep.aggregator.isDefined) {"

  } else {
    interruptibleIter.asInstanceOf[Iterator[(K, Nothing)]]
  }
}
```

The proposed code is effectively this - but spread across two blocks tied through `aggregated` flag.

**Review by mridulm — COMMENTED** [review]
Nice improvement !
Looks good to me, would love to have more eyes on this though.

+CC @Ngone51, @JoshRosen

File: core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala (hunk 1 above)
Comment (zhengchenyu) [reply core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala:128]: Thanks for your review! This does make the code cleaner indeed! I will fix it.

**Comment by mridulm** [comment]
Can you update to latest @zhengchenyu ? That should hopefully fix the build.

Will give a couple of days more for folks to review the patch in meantime.

**Comment by mridulm** [comment]
Merged to master.
Thanks for adding this @zhengchenyu !

**Comment by waitinfuture** [comment]
> > > but combine in ExternalSorter will never trigger extra spill.
> > 
> > 
> > It could be triggered by `ExternalSorter#maybeSpillCollection`?
> 
> @Ngone51 Thanks for your reply!
> 
> Yes, It could be triggered by `ExternalSorter#maybeSpillCollection`. I mean that the Extra spill is trigger by `ExternalAppendOnlyMap#spill`
> 
> Before the change, both`ExternalAppendOnlyMap#spill` and `ExternalSorter#maybeSpillCollection` will be spill. After the change, only `ExternalSorter#maybeSpillCollection` will be spill.
> 
> I created a huge amount of data, and I found the following logs:
> 
> ```
> 23/12/26 16:55:52 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 522.3 MiB to disk (1 time so far)
> 23/12/26 16:56:09 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 629.8 MiB to disk (2 times so far)
> 23/12/26 16:56:40 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 491.5 MiB to disk (3 times so far)
> 23/12/26 16:56:58 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 553.0 MiB to disk (4 times so far)
> 23/12/26 16:57:16 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 442.9 MiB to disk (5 times so far)
> 23/12/26 16:57:47 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 576.1 MiB to disk (6 times so far)
> 23/12/26 16:58:04 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 540.2 MiB to disk (7 times so far)
> 23/12/26 16:58:22 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 686.3 MiB to disk (8 times so far)
> 23/12/26 16:58:52 INFO ExternalAppendOnlyMap: Thread 48 spilling in-memory map of 540.2 MiB to disk (9 times so far)
> 23/12/26 17:00:06 INFO ExternalAppendOnlyMap: Task 5 force spilling in-memory map to disk and it will release 942.2 MiB memory
> 23/12/26 17:00:27 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (1 time so far)
> 23/12/26 17:01:24 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (2 times so far)
> 23/12/26 17:02:24 INFO ExternalSorter: Thread 48 spilling in-memory map of 959.1 MiB to disk (3 times so far)
> 23/12/26 17:03:23 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (4 times so far)
> 23/12/26 17:04:30 INFO ExternalSorter: Thread 48 spilling in-memory map of 948.8 MiB to disk (5 times so far)
> 23/12/26 17:05:24 INFO ExternalSorter: Thread 48 spilling in-memory map of 943.7 MiB to disk (6 times so far)
> 23/12/26 17:06:25 INFO ExternalSorter: Thread 48 spilling in-memory map of 954.0 MiB to disk (7 times so far)
> 23/12/26 17:07:08 INFO ExternalSorter: Thread 48 spilling in-memory map of 959.1 MiB to disk (8 times so far)
> 23/12/26 17:07:58 INFO ExternalSorter: Thread 48 spilling in-memory map of 959.1 MiB to disk (9 times so far) 
> ```
> 
> After the change, we will only found ExternalSorter spill log.

Hi @zhengchenyu , curious is the total number of spilling differ before and after this PR? Thanks!

**Comment by zhengchenyu** [comment]
@waitinfuture 
(1) performance differs
Experimental data were recorded in https://issues.apache.org/jira/browse/SPARK-46512. How much performance improvement depends on the experimental environment. In my experiments, the correlation time was reduced from 75 seconds to 29 seconds.

(2) total number of spilling differs
The reduction in the number of spills is obvious. 
Before this PR, we combine the unsorted records, then sort. When we combine the unsorted records, we us ExternalAppendOnlyMap. They may spill for large data. Then when we sort, we still spill for large data.
After this PR, when we sort, we can easily organize the same keys together, and then we no longer have to use ExternalAppendOnlyMap to combine.
After this PR, we will save the spilled process which we must do in combine.

**Comment by waitinfuture** [comment]
> @waitinfuture (1) performance differs Experimental data were recorded in https://issues.apache.org/jira/browse/SPARK-46512. How much performance improvement depends on the experimental environment. In my experiments, the correlation time was reduced from 75 seconds to 29 seconds.
> 
> (2) total number of spilling differs The reduction in the number of spills is obvious. Before this PR, we combine the unsorted records, then sort. When we combine the unsorted records, we us ExternalAppendOnlyMap. They may spill for large data. Then when we sort, we still spill for large data. After this PR, when we sort, we can easily organize the same keys together, and then we no longer have to use ExternalAppendOnlyMap to combine. After this PR, we will save the spilled process which we must do in combine.

Got it, thanks!

**Pull request closed without merging**